Annotated source code: The Optimism Bridge (#5599)
ethereum/ethereum-org-website: src/content/developers/tutorials/optimism-std-bridge-annotated-code/index.md

     * Cross-chain Functions *
     *************************/

    /**
     * @dev Complete a withdrawal from L2 to L1, and credit funds to the recipient's balance of the
     * L1 ERC20 token.
     * This call will fail if the initialized withdrawal from L2 has not been finalized.
     *
     * @param _l1Token Address of L1 token to finalizeWithdrawal for.
     * @param _l2Token Address of L2 token where withdrawal was initiated.
     * @param _from L2 address initiating the transfer.
     * @param _to L1 address to credit the withdrawal to.
     * @param _amount Amount of the ERC20 to deposit.
     * @param _data Data provided by the sender on L2. This data is provided
     *   solely as a convenience for external contracts. Aside from enforcing a maximum
     *   length, these contracts provide no guarantees about its content.
     */
    function finalizeERC20Withdrawal(
        address _l1Token,
        address _l2Token,
        address _from,
        address _to,
        uint256 _amount,
        bytes calldata _data
    ) external;
}
```

Withdrawals (and other messages from L2 to L1) in Optimism are a two step process:

1. An initiating transaction on L2. 

> 1. A finalizing or claiming transaction on L1.

Reviewer: ```suggestion
2. A finalizing or claiming transaction on L1.
```

            getCrossDomainMessenger().xDomainMessageSender() == _sourceDomainAccount,
            "OVM_XCHAIN: wrong sender of cross-domain message"
        );
```

The way the cross domain messenger provides the address that sent a message with the other layer is [the `.xDomainMessageSender()` function](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts/contracts/L1/messaging/L1CrossDomainMessenger.sol#L122-L128). 
As long as it is called in the transaction that was initiated by the message it can provide this information.

We need to make sure that the message we received came from the other bridge.

```solidity

        _;
    }

    /**********************
     * Internal Functions *
     **********************/

    /**
     * Gets the messenger, usually from storage. This function is exposed in case a child contract
     * needs to override.
     * @return The address of the cross-domain messenger contract which should be used.
     */
    function getCrossDomainMessenger() internal virtual returns (ICrossDomainMessenger) {
        return ICrossDomainMessenger(messenger);
    }
```

This function returns the cross domain messenger. 

> We use a function rather than the variable `messenger` to allow contracts than inherit from this one to use an algorithm to specify which cross domain messenger to use. 

Reviewer 2: ```suggestion
We use a function rather than the variable `messenger` to allow contracts that inherit from this one to use an algorithm to specify which cross domain messenger to use. 
```

```solidity

    /**
     * Sends a message to an account on another domain
     * @param _crossDomainTarget The intended recipient on the destination domain
     * @param _message The data to send to the target (usually calldata to a function with
     *  `onlyFromCrossDomainAccount()`)
     * @param _gasLimit The gasLimit for the receipt of the message on the target domain.
     */
    function sendCrossDomainMessage(
        address _crossDomainTarget,
        uint32 _gasLimit,
        bytes memory _message
```

Finally, the function that sends a message to the other layer. 

```solidity
    ) internal {
        // slither-disable-next-line reentrancy-events, reentrancy-benign
```

[Slither](https://github.com/crytic/slither) is a static analyzer Optimism runs on every contract to look for vulnerabilities and other potential problems.
In this case, the following line triggers two vulnerabilities:

1. [Reentrancy events](https://github.com/crytic/slither/wiki/Detector-Documentation#reentrancy-vulnerabilities-3)

> 1. [Benigh reentrancy](https://github.com/crytic/slither/wiki/Detector-Documentation#reentrancy-vulnerabilities-2)

Reviewer 2: ```suggestion
1. [Benign reentrancy](https://github.com/crytic/slither/wiki/Detector-Documentation#reentrancy-vulnerabilities-2)
```

> 1. [Benign reentrancy](https://github.com/crytic/slither/wiki/Detector-Documentation#reentrancy-vulnerabilities-2)

Reviewer: ```suggestion
2. [Benign reentrancy](https://github.com/crytic/slither/wiki/Detector-Documentation#reentrancy-vulnerabilities-2)
```

```solidity
import { Lib_PredeployAddresses } from "../../libraries/constants/Lib_PredeployAddresses.sol";
```

[`Lib_PredeployAddresses`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts/contracts/libraries/constants/Lib_PredeployAddresses.sol) has the addresses for the L2 contracts that always have the same address. This includes the standard bridge on L2.

```solidity
import { Address } from "@openzeppelin/contracts/utils/Address.sol";
```

[OpenZeppelin's Address utilities](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/Address.sol). It is used to distinguish between contract addresses and those belonging to externally owned accounts (EOA).

Note that this isn't a perfect solution, because there is no way to distinguish between direct calls and calls made from a contract's constructor, but at least this lets us identify and prevent some common user errors.

```solidity
import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
```

[The ERC-20 standard](https://eips.ethereum.org/EIPS/eip-20) supports two ways for a contract to report failure:

1. Revert

> 1. Return `false`

Reviewer: ```suggestion
2. Return `false`
```

```solidity
    function initialize(address _l1messenger, address _l2TokenBridge) public {
        require(messenger == address(0), "Contract has already been initialized.");
```

The `initialize` function should only be called once.
If the address of either the L1 cross domain messenger or the L2 token bridge changes, we create a new proxy and a new bridge that calls it.
This is unlikely to happen except when the entire system is upgraded, a very rare occurrence.

Note that this function does not have any mechanism that restricts *who* can call it.
This means that in theory an attacker could wait until we deploy the proxy and the first version of the bridge and then [front-run](https://solidity-by-example.org/hacks/front-running/) to get to the `initialize` function before the legitimate user does. But there are two methods to prevent this:

1. If the contracts are deployed not directly by an EOA but [in a transaction that has another contract create them](https://medium.com/upstate-interactive/creating-a-contract-with-a-smart-contract-bdb67c5c8595) the entire process can be atomic, and finish before any other transaction is executed.

> 1. If the legitimate call to `initialize` fails it is always possible to ignore the newly created proxy and bridge and create new ones. 

Reviewer: ```suggestion
2. If the legitimate call to `initialize` fails it is always possible to ignore the newly created proxy and bridge and create new ones. 
```

     */
    function _initiateERC20Deposit(
        address _l1Token,
        address _l2Token,
        address _from,
        address _to,
        uint256 _amount,
        uint32 _l2Gas,
        bytes calldata _data
    ) internal {
```

This function is similar to `_initiateETHDeposit` above, with a few important differences.
The first difference is that this function receives the token addresses and the amount to transfer as parameters.
In the case of ETH the call to the bridge already includes the transfer of asset to the bridge acount (`msg.value`).

```solidity
        // When a deposit is initiated on L1, the L1 Bridge transfers the funds to itself for future
        // withdrawals. safeTransferFrom also checks if the contract has code, so this will fail if
        // _from is an EOA or address(0).
        // slither-disable-next-line reentrancy-events, reentrancy-benign
        IERC20(_l1Token).safeTransferFrom(_from, address(this), _amount);
```

ERC-20 token transfers follow a different process from ETH:

1. The user (`_from`) gives an allowance to the bridge to transfer the appropriate tokens.

> 1. The user calls the bridge with the address of the token contract, the amount, etc.

Reviewer: ```suggestion
2. The user calls the bridge with the address of the token contract, the amount, etc.
```

> 1. The bridge transfers the tokens (to itself) as part of the deposit process.

Reviewer: ```suggestion
3. The bridge transfers the tokens (to itself) as part of the deposit process.
```

That contract just doesn't expose them externally, because the conditions to mint and burn tokens are as varied as the number of ways to use ERC-20.

## L2 Bridge Code {#l2-bridge-code}

This is code that runs the bridge on Optimism. 
[The source for this contract is here](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts/contracts/L2/messaging/L2StandardBridge.sol).

```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.9;

/* Interface Imports */
import { IL1StandardBridge } from "../../L1/messaging/IL1StandardBridge.sol";
import { IL1ERC20Bridge } from "../../L1/messaging/IL1ERC20Bridge.sol";
import { IL2ERC20Bridge } from "./IL2ERC20Bridge.sol";
```

The [IL2ERC20Bridge](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts/contracts/L2/messaging/IL2ERC20Bridge.sol) interface is very similar to the [L1 equivalent](#IL1ERC20Bridge) we saw above. 
There are two significant differences:
1. On L1 you initiate deposits and finalize withdrawals. 
   Here you initiate withdrawals and finalize deposits.

> 1. On L1 it is necessary to distinguish between ETH and ERC-20 tokens. 

Reviewer: ```suggestion
2. On L1 it is necessary to distinguish between ETH and ERC-20 tokens. 
```

    }

    /************************************
     * Cross-chain Function: Depositing *
     ************************************/

    /**
     * @inheritdoc IL2ERC20Bridge
     */
    function finalizeDeposit(
        address _l1Token,
        address _l2Token,
        address _from,
        address _to,
        uint256 _amount,
        bytes calldata _data
```

This function is called by `L1StandardBridge`. 

```solidity
    ) external virtual onlyFromCrossDomainAccount(l1TokenBridge) {
```

Make sure the source of the message is legitimate.
This is important because this function calls `_mint` and could be used to give tokens that are not covered by tokens the bridge owns on L1.

```solidity
        // Check the target token is compliant and
        // verify the deposited token on L1 matches the L2 deposited token representation here
        if (
            // slither-disable-next-line reentrancy-events
            ERC165Checker.supportsInterface(_l2Token, 0x1d1d8b63) &&
            _l1Token == IL2StandardERC20(_l2Token).l1Token()
```

Sanity checks:

1. The correct interface is supported

> 1. The L2 ERC-20 contract's L1 address matches the L1 source of the tokens

Reviewer: ```suggestion
2. The L2 ERC-20 contract's L1 address matches the L1 source of the tokens
```

```solidity
        ) {
            // When a deposit is finalized, we credit the account on L2 with the same amount of
            // tokens.
            // slither-disable-next-line reentrancy-events
            IL2StandardERC20(_l2Token).mint(_to, _amount);
            // slither-disable-next-line reentrancy-events
            emit DepositFinalized(_l1Token, _l2Token, _from, _to, _amount, _data);
```

If the sanity checks pass, finalize the deposit:
1. Mint the tokens

> 1. Emit the appropriate event

Reviewer: ```suggestion
2. Emit the appropriate event
```

---
title: "Optimism Standard Bridge Contract Walk-Through"
description: How does the standard bridge for Optimism work? Why does it work this way?
author: Ori Pomerantz
sidebar: true
tags: ["solidity", "optimism", "bridge", "layer 2", "l2"]
skill: intermediate
published: 2022-03-30
lang: en
---

> ## Introduction {#introduction}

Reviewer: ```suggestion
```

[Optimism](https://www.optimism.io/) is an [Optimistic Rollup](/developers/docs/scaling/optimistic-rollups/#top).g

> Optimistic rollups can process transactions for a much lower price than the Ethereum mainnet (also known as layer 1 or L1) because transactions are only processed by a few nodes, instead of the every node on the network. 

Reviewer: ```suggestion
Optimistic rollups can process transactions for a much lower price than Ethereum Mainnet (also known as layer 1 or L1) because transactions are only processed by a few nodes, instead of every node on the network. 
```

> At the same time, the data is all writted to L1 so everything can be proved and reconstructed with all the integrity and availability guarantees of mainnet.

Reviewer: ```suggestion
At the same time, the data is all written to L1 so everything can be proved and reconstructed with all the integrity and availability guarantees of Mainnet.
```

To use L1 assets on Optimism (or any other layer 2, or L2), those assets need to be [bridged](/bridges/#prerequisites).
One way to achieve this is for users to lock assets (ETH and [ERC-20 tokens](https://ethereum.org/en/developers/docs/standards/tokens/erc-20/) are the most common ones) on L1, and receive equivalent assets on L2. 

> Those assets can then be used, transferred, etc. on L2. 

Reviewer: ```suggestion
```

Eventually, whoever ends up with them might want to bridge them back to L1. 
When doing this, the assets are burned on L2 and then released back to the user on L1.

This is the way the [Optimism standard bridge](https://community.optimism.io/docs/developers/bridge/standard-bridge) works. 
In this article we go over the source code for that bridge to see how it works and study it as an example of well written Solidity code.

## Control flows {#control-flows}

The bridge has two main flows:
1. Deposit (from L1 to L2)
1. Withdrawal (from L2 to L1)

### Deposit flow {#deposit-flow}

> #### Layer 1

Reviewer: ```suggestion
#### Layer 1 {#deposit-flow-layer-1}
```

> 1. (ERC-20 only) the depositor gives the bridge an allowance to spend the amount being deposited

Reviewer: ```suggestion
1. If depositing an ERC-20, the depositor gives the bridge an allowance to spend the amount being deposited
```

> 1. The depositor calls the L1 bridge (`depositERC20`, `depositERC20To`, `depositETH`, or `depositETHTo`)

Reviewer: We should number these explicitly to avoid confusion when translating this.

```suggestion
2. The depositor calls the L1 bridge (`depositERC20`, `depositERC20To`, `depositETH`, or `depositETHTo`)
```

> 1. The L1 bridge takes possession of the bridged asset

Reviewer: ```suggestion
3. The L1 bridge takes possession of the bridged asset
```

   - ETH: The asset is transferred by the depositor as part of the call
   - ERC-20: The asset is transferred by the bridge to itself using the allowance provided by the depositor

> 1. The L1 bridge uses the cross domain message mechanism to call `finalizeDeposit` on the L2 bridge

Reviewer: ```suggestion
4. The L1 bridge uses the cross-domain message mechanism to call `finalizeDeposit` on the L2 bridge
```

#### Layer 2

4. The L2 bridge verifies the call to `finalizeDeposit` is legitimate:
   - Came from the cross domain message contract
   - Was originally from the bridge on L1
1. The L2 bridge checks if the ERC-20 token contract on L2 is the correct one:
   - The L2 contract reports that its L1 counterpart is the same as the one the tokens came from on L1
   - The L2 contract reports that it supports the correct interface ([using ERC-165](https://eips.ethereum.org/EIPS/eip-165)).
1. If the L2 contract is the correct one, call it to mint the appropriate number of tokens to the appropriate address. If not, start a withdrawal process to allow the user to claim the tokens on L1.

### Withdrawal flow {#withdrawal-flow}

> #### Layer 2

Reviewer: ```suggestion
#### Layer 2 {#withdrawl-flow-layer-2}
```

1. The withdrawer calls the L2 bridge (`withdraw` or `withdrawTo`)
1. The L2 bridge burns the appropriate number of tokens belonging to `msg.sender`
1. The L2 bridge uses the cross domain message mechanism to call `finalizeETHWithdrawal` or `finalizeERC20Withdrawal` on the L1 bridge 

#### Layer 1

4. The L1 bridge verifies the call to `finalizeETHWithdrawal` or `finalizeERC20Withdrawal` is legitimate:
   - Came from the cross domain message mechanism
   - Was originally from the bridge on L2

> 1. The L1 bridge transfers the appropriate asset (ETH or ERC-20) to the appropriate address

Reviewer: ```suggestion
5. The L1 bridge transfers the appropriate asset (ETH or ERC-20) to the appropriate address
```

## L1 Code {#l1-code}

> This is the code that runs on L1, the mainnet Ethereum network.

Reviewer: ```suggestion
This is the code that runs on L1, the Ethereum Mainnet.
```
